src/calculi/wlp.rs

```diff
@@ -0,0 +1,44 @@
+use std::fmt;
+
+use crate::{
+    ast::{Direction, FileId, Ident, Symbol},
+    proof_rules::EncodingKind,
+};
+
+use super::Calculus;
+
+pub struct WlpCalculus(Ident);
+
+impl WlpCalculus {
+    pub fn new(file: FileId) -> Self {
+        WlpCalculus(Ident::with_dummy_file_span(Symbol::intern("wlp"), file))
+    }
+}
+
+impl fmt::Debug for WlpCalculus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("wlp").finish()
+    }
+}
+
+impl Calculus for WlpCalculus {
+    fn name(&self) -> Ident {
+        self.0
+    }
+
+    fn check_encoding(&self, encoding: EncodingKind, direction: Direction) -> Result<(), ()> {
+        match encoding {
+            EncodingKind::Invariant(_) | EncodingKind::KInduction(_) | EncodingKind::Unroll(_) => {
```

Comment:
These match statements could also be simplified. Instead of nesting every possible case and returning Ok or Err, simply match on the encoding kind and then return the expected direction from the match statement (or immediately abort for Ast, Past, Ost). Then, you can raise an Err if the direction does not match.